lib/containers/k8s.pm

```diff
@@ -17,77 +17,112 @@ use testapi;
+    my $registry = registry_url();
+    assert_script_run "curl " . data_url('containers/registries.yaml') . " -o /etc/rancher/k3s/registries.yaml";
+    file_content_replace("/etc/rancher/k3s/registries.yaml", REGISTRY => $registry);
+}
+
+sub setup_and_check_k3s {
+    set_custom_registry;
+    ensure_k3s_start;
+    check_k3s;
+}
+
 =head2 install_k3s
-Installs k3s, checks the instalation and prepare the kube/config
+Deploy k3s using k3s-install script that is either pulled from upstream or distro
 =cut
 
 sub install_k3s {
-    if (get_var("DISTRI") eq "alp") {
-        assert_script_run('/usr/bin/k3s-install');
-    }
-    else {
-        if (is_microos()) {
-            trup_call('pkg install k3s-selinux');
-            check_reboot_changes;
-        }
-        zypper_call('in apparmor-parser') if (is_sle('>=15-sp1'));
-  # Apply additional options. For more information see https://rancher.com/docs/k3s/latest/en/installation/install-options/#options-for-installation-with-script
-        my $k3s_version = get_var("CONTAINERS_K3S_VERSION");
-        if ($k3s_version) {
-            record_info('k3s forced version', $k3s_version);
-            assert_script_run("export INSTALL_K3S_VERSION=$k3s_version");
+    return if (script_run('which k3s') == 0);
+
+    # Apply additional k3s installation options
+    # Note: The install script starts a k3s-server by default, unless INSTALL_K3S_SKIP_START is set to true
+    # For more information see https://rancher.com/docs/k3s/latest/en/installation/install-options/#options-for-installation-with-script
+    my %k3s_args = (
+        INSTALL_K3S_SYMLINK => get_var('K3S_SYMLINK'),
+        INSTALL_K3S_BIN_DIR => get_var('K3S_BIN_DIR'),
+        INSTALL_K3S_CHANNEL => get_var('K3S_CHANNEL'),
+        INSTALL_K3S_VERSION => get_var("K3S_VERSION"),
+        INSTALL_K3S_SKIP_START => 'true',
+        K3S_NODE_NAME => 'k3s-node'
+    );
+
+    # github.com/k3s-io/k3s#5946 - The kubectl delete namespace helm-ns-413 command freezes and does nothing
+    my $disables = '--disable=metrics-server';
+    $disables .= ' --disable-helm-controller' unless (get_var('K3S_ENABLE_HELM_CONTROLLER'));
+
+    while (my ($key, $value) = each %k3s_args) {
+        if ($value) {
+            assert_script_run("export $key=$value");
         }
-        assert_script_run("export INSTALL_K3S_SYMLINK=" . get_var('K3S_SYMLINK')) if (get_var('K3S_SYMLINK'));
-        assert_script_run("export INSTALL_K3S_BIN_DIR=" . get_var('K3S_BIN_DIR')) if (get_var('K3S_BIN_DIR'));
-        assert_script_run("export INSTALL_K3S_CHANNEL=" . get_var('K3S_CHANNEL')) if (get_var('K3S_CHANNEL'));
-        # k3s doesn't like long hostnames like the ones being used in publiccloud
-        assert_script_run("export K3S_NODE_NAME=k3s-node") if (is_public_cloud);
-
-        my $disables = '--disable=metrics-server';
-        $disables .= ' --disable-helm-controller' unless (get_var('K3S_ENABLE_HELM_CONTROLLER'));
+        delete $k3s_args{$key};
```

Comment:
Isn't it dangerous to remove an item in a dict/map like in Python while iterating on it?

Reply:
https://www.perlmonks.org/?node_id=43716